core/example/solvers/pme.cpp

```diff
@@ -0,0 +1,413 @@
+
+  auto init_p = KOKKOS_LAMBDA( const int idx )
+  {
+    p(idx) = 0.0;
+  };
+  Kokkos::parallel_for( Kokkos::RangePolicy<ExecutionSpace>(0,n_max), init_p ); 
+
+  double alpha = _alpha;
+  //double k_max = _k_max;
+  double r_max = _r_max;
+  double eps_r = _eps_r;
+
+
+#ifdef TDS_CUDA
+  Kokkos::View<int*, Kokkos::CudaUVMSpace> k_max_int("k_max_int",3);
+  for ( auto i = 0; i < 3; ++i)
+  {
+    k_max_int[i] = _k_max_int[i];
+  }
+#else
+  int* k_max_int = &(_k_max_int[0]);
+#endif
+  std::cout << "Starting real-space contribution" << std::endl;
+  struct timeval starttime;
+  gettimeofday(&starttime, NULL);
+  // computation real-space contribution
+  Kokkos::parallel_reduce( Kokkos::RangePolicy<ExecutionSpace>(0,n_max), KOKKOS_LAMBDA(int idx, double& Ur_part)
+      {
+        double d[SPACE_DIM];
+        double k; 
+          for (auto i = 0; i < n_max; ++i)
+          {
+            for (auto j = 0; j < 3; ++j)
+              d[j] = r( idx, j ) - r( i, j );
+            double qiqj = q( idx ) * q( i );
+            for (auto kx = 0; kx <= k_max_int[0]; ++kx)
+            {
+              k = (double)kx * lx;
+              if (k - lx > r_max) continue;
+              for (auto ky = 0; ky <= k_max_int[1]; ++ky)
+              {
+                k = sqrt( (double)kx * (double)kx * lx * lx +
+                  (double)ky * (double)ky * ly * ly);
+                if (k - lx  > r_max) continue;
+                for (auto kz = 0; kz <= k_max_int[2]; ++kz)
+                {
+                  if ( kx == 0 && ky == 0 && kz == 0 && i == idx) continue;
+
+                  k = sqrt( (double)kx * (double)kx * lx * lx +
+                            (double)ky * (double)ky * ly * ly +
+                            (double)kz * (double)kz * lz * lz );
+                  if (k - lx > r_max) continue;
+
+                  double scal = (d[0] + (double)kx * lx) * (d[0] + (double)kx * lx) +
+                    (d[1] + (double)ky * ly) * (d[1] + (double)ky * ly) +
+                    (d[2] + (double)kz * lz) * (d[2] + (double)kz * lz);
+                  scal = sqrt(scal);
+                  if (scal > r_max) 
+                    continue;
+                  Ur_part += qiqj * erfc(alpha * scal)/scal;
+                }
+              }
+            }
+          }
+        Ur_part *= 0.5;
+        p(idx) += Ur_part;
+      },
+    Ur
+  );
+
+  std::cout << "End of real-space contribution" << std::endl;
+  
+  // computation reciprocal-space contribution
+ 
+  //First, spread the charges onto the mesh
+  std::cout << "spreading particle charge to mesh" << std::endl;
+  
+  double spacing = meshr(1,0)-meshr(0,0);//how far apart the mesh points are (assumed uniform cubic)   
+  
+  int n_max_mesh = mesh.size();
+  auto spread_q = KOKKOS_LAMBDA( const int idx )
+  {
+     double xdist, ydist, zdist; //could make this an array
+     //Find which particles are within range to add charge
+     for ( size_t pidx = 0; pidx < particles.size(); ++pidx )
```

Comment:
Yes, this is a very interesting point. @rhalver and I discussed that different methods of parallelism here _may_ perform better depending on some factors: GPU/CPU, number of particles vs mesh points, ... 

It could be an interesting test at some point to try both particle-centric spreading of charge to mesh points, and mesh-centric gathering of charge.